File: frontend/src/api/eda.js
/**
 * EDA API calls.
 */

import { API_BASE_URL } from './client.js';
import { getAccessToken } from './tokenStorage.js';

function authHeaders(json = true) {
  const headers = { Accept: 'application/json' };
  if (json) {
    headers['Content-Type'] = 'application/json';
  }
  const token = getAccessToken();
  if (token) {
    headers.Authorization = `Bearer ${token}`;
  }
  return headers;
}

/**
 * Run EDA on an uploaded file.
 *
 * @param {number} fileId
 * @param {{ forceRefresh?: boolean }} [options]
 */
export async function runEda(fileId, options = {}) {
  const response = await fetch(`${API_BASE_URL}/eda/${fileId}`, {
    method: 'POST',
    headers: authHeaders(),
    body: JSON.stringify({
      force_refresh: options.forceRefresh ?? false,
      force_background: options.forceBackground ?? false,
    }),
  });
  const body = await response.json().catch(() => ({}));
  if (response.status === 202) {
    return body;
  }
  if (!response.ok) {
    throw new Error(body.message ?? body.detail ?? 'EDA analysis failed');
  }
  return body;
}

/**
 * Fetch cached EDA results.
 *
 * @param {number} fileId
 */
export async function getEda(fileId) {
  const response = await fetch(`${API_BASE_URL}/eda/${fileId}`, {
    headers: authHeaders(false),
  });
  if (response.status === 204) {
    return null;
  }
  if (!response.ok) {
    const body = await response.json().catch(() => ({}));
    throw new Error(body.message ?? body.detail ?? 'EDA results not found');
  }
  return response.json();
}

/**
 * Fetch modelling suggestions from EDA.
 *
 * @param {number} fileId
 */
export async function getEdaSuggestions(fileId) {
  const response = await fetch(`${API_BASE_URL}/eda/${fileId}/suggestions`, {
    headers: authHeaders(false),
  });
  if (!response.ok) {
    const body = await response.json().catch(() => ({}));
    throw new Error(body.message ?? body.detail ?? 'Failed to load EDA suggestions');
  }
  return response.json();
}

File: backend/app/routers/eda.py
"""EDA API routes.

Exposes exploratory data analysis endpoints for uploaded CSV files.
"""

from typing import Annotated

from fastapi import APIRouter, Depends, HTTPException, status
from fastapi.responses import Response
from sqlalchemy.orm import Session

from app.auth.dependencies import require_analyst, require_viewer
from app.database import get_db
from app.models.user import User
from app.schemas.eda import (
    ColumnSummary,
    EDARequest,
    EDAResponse,
    EDASuggestions,
)
from app.services.eda_service import EDAError, eda_service

router = APIRouter(prefix="/eda", tags=["eda"])


@router.post(
    "/{file_id}",
    response_model=EDAResponse,
    summary="Run EDA on an uploaded file",
)
def run_eda(
    file_id: int,
    request: EDARequest,
    db: Annotated[Session, Depends(get_db)],
    _: Annotated[User, Depends(require_analyst)],
) -> EDAResponse:
    """Run or return cached EDA for a stored CSV file."""
    try:
        return eda_service.run_for_file(
            db,
            file_id=file_id,
            force_refresh=request.force_refresh,
        )
    except EDAError as exc:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=str(exc),
        ) from exc


@router.get(
    "/{file_id}",
    response_model=EDAResponse,
    summary="Get cached EDA results",
    responses={204: {"description": "EDA has not been run for this file yet"}},
)
def get_eda(
    file_id: int,
    db: Annotated[Session, Depends(get_db)],
    _: Annotated[User, Depends(require_viewer)],
) -> EDAResponse | Response:
    """Return cached EDA output for a file."""
    try:
        return eda_service.get_cached_for_file(db, file_id=file_id)
    except EDAError as exc:
        if "not been run" in str(exc).lower():
            return Response(status_code=status.HTTP_204_NO_CONTENT)
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=str(exc),
        ) from exc


@router.get(
    "/{file_id}/columns/{column_name}",
    response_model=ColumnSummary,
    summary="Get EDA summary for one column",
)
def get_column_eda(
    file_id: int,
    column_name: str,
    db: Annotated[Session, Depends(get_db)],
    _: Annotated[User, Depends(require_viewer)],
) -> ColumnSummary:
    """Return EDA details for a single column."""
    try:
        return eda_service.get_column_summary(
            db,
            file_id=file_id,
            column_name=column_name,
        )
    except EDAError as exc:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=str(exc),
        ) from exc


@router.get(
    "/{file_id}/suggestions",
    response_model=EDASuggestions,
    summary="Get modelling suggestions from EDA",
)
def get_eda_suggestions(
    file_id: int,
    db: Annotated[Session, Depends(get_db)],
    _: Annotated[User, Depends(require_viewer)],
) -> EDASuggestions:
    """Return suggested targets, features, and analyses."""
    try:
        return eda_service.get_suggestions(db, file_id=file_id)
    except EDAError as exc:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=str(exc),
        ) from exc
